be/src/storage/compaction_manager.cpp

```diff
@@ -561,8 +560,32 @@ std::unordered_set<CompactionTask*> CompactionManager::get_running_task(const Ta
     return res;
 }
 
+int32_t CompactionManager::compute_max_compaction_concurrency() const {
+    int32_t max_task_num = 0;
+    // new compaction framework
+    if (config::base_compaction_num_threads_per_disk >= 0 && config::cumulative_compaction_num_threads_per_disk >= 0) {
+        max_task_num = static_cast<int32_t>(
+                StorageEngine::instance()->get_store_num() *
+                (config::cumulative_compaction_num_threads_per_disk + config::base_compaction_num_threads_per_disk));
+    } else {
+        // When cumulative_compaction_num_threads_per_disk or config::base_compaction_num_threads_per_disk is less than 0,
+        // there is no limit to _max_task_num if max_compaction_concurrency is also less than 0, and here we set maximum value to be 20.
+        max_task_num = std::min(20, static_cast<int32_t>(StorageEngine::instance()->get_store_num() * 5));
+    }
+    if (config::max_compaction_concurrency > 0 && config::max_compaction_concurrency < max_task_num) {
+        max_task_num = config::max_compaction_concurrency;
+    }
+
+    return max_task_num;
+}
+
 Status CompactionManager::update_max_threads(int max_threads) {
     if (_compaction_pool != nullptr) {
+        if (max_threads == -1) {
+            std::lock_guard lg(_tasks_mutex);
```

Comment:
reminder me again why this lock is required? protect the modification of _max_task_num?

is it ok to run the update_max_threads() out of the lock scope?

Reply:
Sorry, I used the wrong lock.